### PR TITLE
curvefs improve the inode and dentry cache lock

### PR DESCRIPTION
**Pull request by SeanHai**
### What problem does this PR solve?

Issue Number: close #xxx 
#810 
Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by SeanHai** [comment]
recheck

File: curvefs/src/client/inode_cache_manager.cpp [hunk 1]
```diff
@@ -44,12 +44,6 @@ CURVEFS_ERROR InodeCacheManagerImpl::GetInode(uint64_t inodeid,
...
     }
 
-    curve::common::WriteLockGuard lg(mtx_);
-    bool ok = iCache_->Get(inodeid, &out);
-    if (ok) {
-        return CURVEFS_ERROR::OK;
-    }
-
     Inode inode;
     MetaStatusCode ret2 = metaClient_->GetInode(fsId_, inodeid, &inode);
```
Comment (xu-chaojie) [inline curvefs/src/client/inode_cache_manager.cpp:48]: if don't lock GetInode here, when update inode between two GetInode was called at the same time,  maybe second GetInode will return the old version inode？

Comment (xu-chaojie) [inline curvefs/src/client/inode_cache_manager.cpp:48]: If not lock GetInode here， when an update inode and read inode concuently call， the read inode maybe return old version of inode

File: curvefs/src/client/dentry_cache_manager.cpp [hunk 2]
```diff
@@ -61,11 +59,6 @@ CURVEFS_ERROR DentryCacheManagerImpl::GetDentry(uint64_t parent,
             return CURVEFS_ERROR::OK;
         }
     }
-    curve::common::WriteLockGuard lg(mtx_);
-    bool ok = dCache_->Get(key, out);
-    if (ok) {
-        return CURVEFS_ERROR::OK;
-    }
 
     MetaStatusCode ret = metaClient_->GetDentry(fsId_, parent, name, out);
```
Comment (xu-chaojie) [inline curvefs/src/client/dentry_cache_manager.cpp:63]: if don‘t lock GetDentry here， when concurrent with InsertOrReplaceCache above， this fuction maybe return old version of dentry？

File: curvefs/src/client/inode_cache_manager.cpp [hunk 3]
```diff
@@ -64,7 +58,12 @@ CURVEFS_ERROR InodeCacheManagerImpl::GetInode(uint64_t inodeid,
         std::move(inode), metaClient_);
 
     std::shared_ptr<InodeWrapper> eliminatedOne;
-    bool eliminated = iCache_->Put(inodeid, out, &eliminatedOne);
+    bool eliminated;
```
Comment (ilixiaocui) [inline curvefs/src/client/inode_cache_manager.cpp:61]: eliminated need to be initialized

File: curvefs/src/client/inode_cache_manager.cpp [hunk 4]
```diff
@@ -91,7 +89,11 @@ CURVEFS_ERROR InodeCacheManagerImpl::CreateInode(
...
```
Comment (ilixiaocui) [inline curvefs/src/client/inode_cache_manager.cpp:92]: ditto

File: curvefs/src/client/dentry_cache_manager.cpp (hunk 2 above)
Comment (ilixiaocui) [reply curvefs/src/client/dentry_cache_manager.cpp:63]: we need get from cache first, if cache do not have this one, then  get from metaserver.

may be readlock can be use by get.

Comment (SeanHai) [reply curvefs/src/client/dentry_cache_manager.cpp:63]: Using namelock to lock the specific one. If lock entire icache will block requests to other inode.

Comment (SeanHai) [reply curvefs/src/client/dentry_cache_manager.cpp:63]: > if don‘t lock GetDentry here， when concurrent with InsertOrReplaceCache above， this fuction maybe return old version of dentry？

Using namelock to lock the specific one. If lock entire icache will block requests to other inode.

Comment (SeanHai) [reply curvefs/src/client/dentry_cache_manager.cpp:63]: > we need get from cache first, if cache do not have this one, then get from metaserver.
> 
> may be readlock can be use by get.

The key issue is the lock for dCache will block requests on other dentry. Using namelock to lock the specific one.

File: curvefs/src/client/inode_cache_manager.cpp (hunk 3 above)
Comment (SeanHai) [reply curvefs/src/client/inode_cache_manager.cpp:61]: > eliminated need to be initialized

fixed

File: curvefs/src/client/inode_cache_manager.cpp (hunk 4 above)
Comment (SeanHai) [reply curvefs/src/client/inode_cache_manager.cpp:92]: > ditto

fixed

**Comment by SeanHai** [comment]
recheck

**Review by ilixiaocui — APPROVED** [review]

**Comment by SeanHai** [comment]
recheck

**Comment by SeanHai** [comment]
recheck

File: curvefs/src/client/dentry_cache_manager.cpp [hunk 5]
```diff
@@ -61,10 +60,14 @@ CURVEFS_ERROR DentryCacheManagerImpl::GetDentry(uint64_t parent,
...
     }
-    curve::common::WriteLockGuard lg(mtx_);
-    bool ok = dCache_->Get(key, out);
-    if (ok) {
-        return CURVEFS_ERROR::OK;
+
+    NameLockGuard lock(nameLock_, key);
+    {
+        curve::common::ReadLockGuard lg(mtx_);
+        bool ok = dCache_->Get(key, out);
```
Comment (xu-chaojie) [inline curvefs/src/client/dentry_cache_manager.cpp:67]: LRU cache already lock put & get in its own logical，maybe mtx_ here is not needed？

Comment (SeanHai) [reply curvefs/src/client/dentry_cache_manager.cpp]: > LRU cache already lock put & get in its own logical，maybe mtx_ here is not needed？

Yes, the lock into LRU can protect operations on cache, the namelock can sequence the operations on the same one.

**Review by xu-chaojie — APPROVED** [review]

**Pull request merged**